Automattic/newspack-blocks
# feat: checkout button block

### All Submissions:






### Changes proposed in this Pull Request:

Implements a new "Checkout Button" block to provide a quick modal checkout of a WooCommerce product.

To achieve that the logic behind the Donate block modal checkout has been detached and abstracted to a separate class so it can be used by both blocks.

The [cart generation for the Donate block](https://github.com/Automattic/newspack-plugin/blob/336ee942f31cba9613f2c3065c7475165ee5fa9d/includes/class-donations.php#L586-L705) has been preserved and inspired a [new method implemented for the more abstract cart generation](https://github.com/Automattic/newspack-blocks/blob/ca54269b88b13507e9e82853094a842e957ba51d/includes/class-modal-checkout.php#L37-L130).

While testing you'll notice that any custom selection of billing fields in the "Reader Revenue" wizard will not apply to custom products used in the block. That's because this custom selection applies only to "Donate" products.

The block also supports applying a custom price to the checkout of the selected product if it has "Name Your Price".

<img width="938" alt="image" src="https://user-images.githubusercontent.com/820752/230482303-3dff6725-1e19-44b3-9be6-f4b07fe40b38.png">

<img width="915" alt="image" src="https://user-images.githubusercontent.com/820752/230482400-0c8e7f06-508b-4bcc-8f3c-23daaac8bf3e.png">

### How to test the changes in this Pull Request:

1. Make sure you have different Woo products (including with Name Your Price)
2. Checkout this branch, edit a page, and include the new "Checkout Button" block
3. Add a text to the block and search for a product
4. Save the page and visit
5. Click on the button, confirm the modal checkout flow starts and you are able to complete the purchase
6. Edit the page and select a product with Name Your Price
7. Confirm you see the input for the custom price
8. Change the price, save, and repeat step 5 confirming the custom price is applied

#### Styling

1. Tweak the styles of the button and confirm they are applied on the editor and on the frontend
2. Change the alignment to center and full and confirm it's also properly applied

<img width="946" alt="image" src="https://user-images.githubusercontent.com/820752/230483741-14153b88-b53d-443b-bc25-eeaa61abeef1.png">

### Other information:
